### PR TITLE
Breestryker/armdispatch

### PR DESCRIPTION
**Pull request by Breanna-Stryker**
# Description

Code change adds a template and a wrapper script to deploy MLZ via a quickstart button.  Options also allow for packaging as a marketplace app.

**NOTE**  These changes are not testable while in branch.  URI's in scripts are pointing to their post PR change locations. 
If you wish to test you can change the URI's in the readme to:
https%3A%2F%2Fraw.githubusercontent.com%2FAzure%2Fmissionlz%2Fbreestryker%2Farmdispatch%2Fsrc%2Fbuild%2Farm_quickstart%2FmainTemplate.json

And the mainTemplate.json primaryScriptUrl must be changed to:
https://raw.githubusercontent.com/Azure/missionlz/breestryker/armdispatch/src/build/arm_quickstart/arm_deploy.sh

## Issue reference

The issue this PR will close: #260 

## Checklist

Please make sure you've  completed the relevant tasks for this PR, out of the following list:

* [X] Code compiles or validates correctly
* [X] BASH scripts have been validated using `shellcheck`
* [X] All tests pass (manual and automated)
* [X] The documentation is updated to cover any new or changed features
* [X] Markdown files have been linted using the recommended linter. (See `.vscode/extensions.json`.)
* [X] Relevant issues are linked to this PR

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -2,6 +2,9 @@
 
 Mission Landing Zone is a highly opinionated template which IT oversight organizations can use to create a cloud management system to deploy Azure environments for their teams. It addresses a narrowly scoped, specific need for an SCCA compliant hub and spoke infrastructure.
 
+[![Deploy to Azure](https://aka.ms/deploytoazurebutton)](https://portal.azure.com/#create/Microsoft.Template/uri/https%3A%2F%2Fraw.githubusercontent.com%2FAzure%2Fmissionlz%2Fmain%2Fsrc%2Fbuild%2Farm_quickstart%2FmainTemplate.json)
+[![Deploy to Azure](https://aka.ms/deploytoazuregovbutton)](https://portal.azure.us/#create/Microsoft.Template/uri/https%3A%2F%2Fraw.githubusercontent.com%2FAzure%2Fmissionlz%2Fmain%2Fsrc%2Fbuild%2Farm_quickstart%2FmainTemplate.jsonn)
```
Comment (glennmusa) [inline README.md:6]: extra char

```suggestion
[![Deploy to Azure](https://aka.ms/deploytoazuregovbutton)](https://portal.azure.us/#create/Microsoft.Template/uri/https%3A%2F%2Fraw.githubusercontent.com%2FAzure%2Fmissionlz%2Fmain%2Fsrc%2Fbuild%2Farm_quickstart%2FmainTemplate.json)
```

File: src/build/arm_quickstart/mainTemplate.json [hunk 2]
```diff
@@ -0,0 +1,123 @@
+{
+    "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#",
+    "contentVersion": "1.0.0.0",
+    "parameters": {
+      "service_principal_id": {
```
Comment (glennmusa) [inline src/build/arm_quickstart/mainTemplate.json:5]:
```suggestion
      "servicePrincipalId": {
```

For this variable, and the others in this template, [ARM templates prefer camelCasing](https://docs.microsoft.com/en-us/azure/azure-resource-manager/templates/best-practices#general-recommendations-for-parameters) and will generate space separated values in the UI and other tools when interpreting them, e.g. `servicePrincipalId` will render `Service Principal ID`

File: src/build/arm_quickstart/mainTemplate.json [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+          "description": "If any value is present, will not create Bastion Host and Jumpbox VM"
+        }
+      },
+      "no_sentinel": {
+        "type": "string",
+        "defaultValue": "true",
+        "metadata": {
+          "description": "If any value is present, will not create an Azure Sentinel solution"
+        }
+      },
```
Comment (glennmusa) [inline src/build/arm_quickstart/mainTemplate.json:58]:
```suggestion
      "no_bastion": {
        "type": "bool",
        "defaultValue": "true",
        "metadata": {
          "description": "If any value is present, will not create Bastion Host and Jumpbox VM"
        }
      },
      "no_sentinel": {
        "type": "bool",
        "defaultValue": "true",
        "metadata": {
          "description": "If any value is present, will not create an Azure Sentinel solution"
        }
      },
```

We could use the [bool type](https://docs.microsoft.com/en-us/azure/azure-resource-manager/templates/data-types#booleans) and the UI will render a radio button-ish component.

File: src/build/arm_quickstart/arm_deploy.sh [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+# Assemble argument list
+az login --service-principal \
+--user "${ARM_CLIENT_ID}" \
+--password "${ARM_CLIENT_SECRET}" \
+--tenant "${env_tenant_id}" \
+--allow-no-subscriptions \
+--output json
+
+./deploy.sh -s "${subscription_id}" -l "${mlz_location}" -e "${tf_environment}" -z "${mlz_env_name}" -w --no-bastion --no-sentinel --no-subscription-id
```
Comment (glennmusa) [inline src/build/arm_quickstart/arm_deploy.sh:34]:
```suggestion
./deploy.sh -s "${subscription_id}" -l "${mlz_location}" -e "${tf_environment}" -z "${mlz_env_name}" -w --no-bastion --no-sentinel
```

Deployment fails on `--no-subscription-id` as an unexpected argument. The output from my deployment:

```json
{
  "code": "DeploymentScriptError",
  "message": "ERROR: Unexpected argument: --no-subscription-id"
}
```

File: README.md [hunk 5]
```diff
@@ -2,6 +2,9 @@
 
 Mission Landing Zone is a highly opinionated template which IT oversight organizations can use to create a cloud management system to deploy Azure environments for their teams. It addresses a narrowly scoped, specific need for an SCCA compliant hub and spoke infrastructure.
 
+[![Deploy to Azure](https://aka.ms/deploytoazurebutton)](https://portal.azure.com/#create/Microsoft.Template/uri/https%3A%2F%2Fraw.githubusercontent.com%2FAzure%2Fmissionlz%2Fmain%2Fsrc%2Fbuild%2Farm_quickstart%2FmainTemplate.json)
+[![Deploy to Azure](https://aka.ms/deploytoazuregovbutton)](https://portal.azure.us/#create/Microsoft.Template/uri/https%3A%2F%2Fraw.githubusercontent.com%2FAzure%2Fmissionlz%2Fmain%2Fsrc%2Fbuild%2Farm_quickstart%2FmainTemplate.json)
+
```
Comment (glennmusa) [inline README.md:7]: Think we could use some end-user documentation in a section like 'ARM Template Deployment' or similar that outlines the pre-requisites (the Service Principal with Contributor RBAC role), how this works (what happens when you click 'Create' from the Portal), and what to expect.

File: src/build/arm_quickstart/arm_deploy.sh (hunk 4 above)
Comment (glennmusa) [reply src/build/arm_quickstart/arm_deploy.sh:34]: This was fixed by merging main

File: .devcontainer/Dockerfile [hunk 6]
```diff
@@ -6,15 +6,15 @@ FROM ubuntu:20.04
 ENV DEBIAN_FRONTEND=noninteractive
 
 # Terraform, providers and tflint versions
-ARG TERRAFORM_VERSION=1.0.3
-ARG AZURERM_VERSION=2.69.0
+ARG TERRAFORM_VERSION=1.0.0
```
Comment (glennmusa) [inline .devcontainer/Dockerfile:9]: The TF version is pinned to `= 1.04` in main.tf and tier3.tf

File: .devcontainer/Dockerfile [hunk 7]
```diff
@@ -6,15 +6,15 @@ FROM ubuntu:20.04
...
+ARG AZURERM_VERSION=2.67.0
```
Comment (glennmusa) [inline .devcontainer/Dockerfile:10]: `azurerm` is pinned to `= 2.70.0`

File: README.md (hunk 5 above)
Comment (glennmusa) [reply README.md:7]: Nice doc update, thanks!

File: .devcontainer/Dockerfile (hunk 6 above)
Comment (Breanna-Stryker) [reply .devcontainer/Dockerfile]: Fixed!

File: .devcontainer/Dockerfile (hunk 7 above)
Comment (Breanna-Stryker) [reply .devcontainer/Dockerfile]: Fixed!

**Pull request closed without merging**